### ISSUE TITLE
ImageNet wrong label encoding and download link dead

### ISSUE THREAD

**Issue by lsmaerz** [giughi/An-Empirical-Study-of-DFO-Algorithms-for-Targeted-Black-Box-Attacks-in-DNNs#2]
The provided download link `http://jaina.cs.ucdavis.edu/datasets/adv/imagenet/img.tar.gz` for ImageNet is dead. I have tried downloading the offical 2011 ImageNet validation set directly from `image-net.org` but this seems to be in a different folder/naming structure causing errors when running `python Setups/Attacks.py --dataset=ImageNet --attack=boby --eps=0.1 --test_size=1 --Adversary_trained=False`.

I am grateful for your suggestions.

**Comment by lsmaerz** [comment]
In particular: Which generation (2011,...,2017) and which part (train, val, test) of the ImageNet dataset was used for the plots? I keep getting ground truth discrepancies.

**Comment by giughi** [comment]
Hi Lars,
sorry for the late reply but I'm having a crazy period at work and cannot
really get back into this unfortunately.
The results should be quite general to any version of ImageNet and should
converge for a large enough number of attacks.
Would you mind sharing what kind of discrepancy you're seeing?
Best,
Giuseppe

**Comment by lsmaerz** [comment]
Hey Giuseppe,
I'm sorry to hear that, hope you get some time to breathe soon!
The `readimg` function in the `setup_inception(_2)` file seems to read the ground truth label from the image file name. Therefore I downloaded the ImageNet2012 validation dataset with its development kit to attach the labels to the image file names. The resulting groups seem reasonable acc. to the dev kit and `https://image-net.org/challenges/LSVRC/2012/browse-synsets.php`. 

However, when running `python Setups/Attacks.py --dataset=ImageNet --attack=boby --eps=0.12 --test_size=10 --Adversary_trained=False`, I get outputs of the form `skip wrongly classified image no. 0, original class 34, classified as 276`. Note that 34 = hyena in the official dev kit but also 276 = hyena in `https://deeplearning.cms.waikato.ac.nz/user-guide/class-maps/IMAGENET/`. So why does the model predict according to the last link and what kind of standard is this?
Thanks

**Comment by giughi** [comment]
Hey Lars,
unfortunately I haven't worked anymore on adversarial attacks since the
article and lost touch with the ImageNet development. Though my
understanding from what you wrote is that most probably the inception v3
net is tuned on the old dataset that I was using and that the indexes have
been changed.
If the aim is not to reproduce exactly my results I would probably suggest
refactoring the code to import a new version of the inception v3 net and
that should be consistent with the dataset.
I'm sorry not to be more helpful/
All the best,
Giuseppe

**Comment by lsmaerz** [comment]
Hey Giuseppe,
After downloading the 2012 validation dataset (you also used the 2012 version) from `https://www.image-net.org/`, I was able to fix the problem by translating the different label encodings and prepending the ground truth labels to the validation images. This is my handcrafted translation file `trans_row.txt`
[trans_row.txt](https://github.com/giughi/An-Empirical-Study-of-DFO-Algorithms-for-Targeted-Black-Box-Attacks-in-DNNs/files/12551036/trans_row.txt)
located in `Setups`. To execute the renaming - either to the correct or the modified labels, I added two files to `Setups/Data_and_Model`, namely:
`import os
def attach_text_to_filenames(txt_file_path, trans_file_path, dir_path):
    with open(txt_file_path) as f:
        text = f.readlines()
    with open(trans_file_path) as t:
        trans = t.read()
    trans.replace('\n', '')
    tr = trans.split(';')
    files = [f for f in os.listdir(dir_path) if os.path.isfile(os.path.join(dir_path, f))]
    files.sort()
    for i, filename in enumerate(files):
        if i >= len(text):
            break
        filepath = os.path.join(dir_path, filename)
        new_filename = tr[int(text[int(filename.split("_")[2].split(".")[0])-1].strip())-1] + '_' + filename
        new_filepath = os.path.join(dir_path, new_filename)
        print(new_filepath)
        os.rename(filepath, new_filepath)

if __name__=="__main__":
    attach_text_to_filenames('./Data/ImageNet/ILSVRC2012_devkit_t12/data/ILSVRC2012_validation_ground_truth.txt', './Data/ImageNet/trans_row.txt', './Data/ImageNet/images')`
for the modified ones and
`import os
def attach_text_to_filenames(txt_file_path, dir_path):
    with open(txt_file_path) as f:
        text = f.readlines()
    files = [f for f in os.listdir(dir_path) if os.path.isfile(os.path.join(dir_path, f))]
    files.sort()
    for i, filename in enumerate(files):
        if i >= len(text):
            break
        filepath = os.path.join(dir_path, filename)
        new_filename = str(int(text[int(filename.split("_")[2].split(".")[0])-1])-1) + '_' + filename
        new_filepath = os.path.join(dir_path, new_filename)
        print(new_filepath)
        os.rename(filepath, new_filepath)

if __name__=="__main__":
    attach_text_to_filenames('./Data/ImageNet/ILSVRC2012_devkit_t12/data/ILSVRC2012_validation_ground_truth.txt', './Data/ImageNet/images')`
for the original one. Hope that helps future users.

Best,
Lars

**Comment by lsmaerz** [comment]
Hey Giuseppe,
After downloading the 2012 validation dataset (you also used the 2012 version) from `https://www.image-net.org/`, I was able to fix the problem by translating the different label encodings and prepending the ground truth labels to the validation images. This is my handcrafted translation file `trans_row.txt`
[trans_row.txt](https://github.com/giughi/An-Empirical-Study-of-DFO-Algorithms-for-Targeted-Black-Box-Attacks-in-DNNs/files/12551036/trans_row.txt)
located in `Setups`. To execute the renaming - either to the correct or the modified labels, I added two files to `Setups/Data_and_Model`, namely:
```
import os
def attach_text_to_filenames(txt_file_path, trans_file_path, dir_path):
    with open(txt_file_path) as f:
        text = f.readlines()
    with open(trans_file_path) as t:
        trans = t.read()
    trans.replace('\n', '')
    tr = trans.split(';')
    files = [f for f in os.listdir(dir_path) if os.path.isfile(os.path.join(dir_path, f))]
    files.sort()
    for i, filename in enumerate(files):
        if i >= len(text):
            break
        filepath = os.path.join(dir_path, filename)
        new_filename = tr[int(text[int(filename.split("_")[2].split(".")[0])-1].strip())-1] + '_' + filename
        new_filepath = os.path.join(dir_path, new_filename)
        print(new_filepath)
        os.rename(filepath, new_filepath)

if __name__=="__main__":
    attach_text_to_filenames('./Data/ImageNet/ILSVRC2012_devkit_t12/data/ILSVRC2012_validation_ground_truth.txt', './Data/ImageNet/trans_row.txt', './Data/ImageNet/images')
```
for the modified ones and
```
import os
def attach_text_to_filenames(txt_file_path, dir_path):
    with open(txt_file_path) as f:
        text = f.readlines()
    files = [f for f in os.listdir(dir_path) if os.path.isfile(os.path.join(dir_path, f))]
    files.sort()
    for i, filename in enumerate(files):
        if i >= len(text):
            break
        filepath = os.path.join(dir_path, filename)
        new_filename = str(int(text[int(filename.split("_")[2].split(".")[0])-1])-1) + '_' + filename
        new_filepath = os.path.join(dir_path, new_filename)
        print(new_filepath)
        os.rename(filepath, new_filepath)

if __name__=="__main__":
    attach_text_to_filenames('./Data/ImageNet/ILSVRC2012_devkit_t12/data/ILSVRC2012_validation_ground_truth.txt', './Data/ImageNet/images')
```
for the original one. Hope that helps future users.

Best,
Lars
